Add home page with articles list

timriley opened the pull request:
This PR adds a home page with a date-descending list of published articles.

To do this, we've put in place all the basics to establish articles as an entity within our system:

- A database migration to create the table
- A ROM relation to model the table
- A repository in the "main" sub-app for returning a list of published articles, date-descending

We've also created an "home" view controller which uses the repository to pass the list of articles to the template, with each article value being wrapped up in a view part class that adds support for rendering the body markdown as HTML.

We've added tests for each of these elements, too. Firstly, we defined a test factory for generating articles (either as simple structs or records in the database), then we added unit tests for the view and view part, as well as a single integration/feature spec to test for the correct homepage being served up via a web request.

timriley committed:
Update gems

timriley committed:
Create articles table

timriley committed:
Add a test factory for articles

timriley committed:
Test for articles showing on homepage, ordered date descending

timriley committed:
Add articles repo (with home_page_listing method) and ROM relation

timriley committed:
Add home page view and render list of article

solnic commented on db/migrate/20170711235422_create_articles.rb:
```diff
@@ -0,0 +1,14 @@
+      column :slug, :text, null: false, unique: true
+      column :title, :text, null: false
+      column :body, :text, null: false
+      column :published, :boolean, default: false, null: false
```
It'd be worth making this one an index and it'd result in `Articles#published` auto-defined for you.

timriley replied:
@solnic Oh yes, great suggestion. I've just made this change in https://github.com/dry-rb/dry-web-blog/pull/1/commits/9364dfa22c3caa78b2eb61a5ef5f2c52f999f372, but I note there's no `Articles#published`, but rather `Articles#by_published` – is that what you meant?

solnic replied:
@timriley gah sorry, yeah it'd be `by_published(value)` doing `where(self[:published] => value)` which in this particular case doesn't look good

timriley committed:
Index articles columns, use auto-generated by_published relation method

choonkeat commented on db/migrate/20170711235422_create_articles.rb:
```diff
@@ -0,0 +1,14 @@
+      column :updated_at, :timestamp, null: false, default: Sequel.lit("(now() at time zone 'utc')")
+    end
+  end
+end
```
minor, but for a rom-rb/dry-web newbie like me following along, it would help to update `README.md` to mention `rake db:migrate` and `rake spec`

``` diff
diff --git a/README.md b/README.md
index 86f9e09..534ca31 100644
--- a/README.md
+++ b/README.md
@@ -24,4 +24,11 @@ To run the app, follow these steps:
 2. Copy `.example.env` to `.env`
 3. Copy `.example.env.test` to `.env.test`
 4. Create a `dry_web_blog_development` Postgres database
-5. Run the app with `bundle exec shotgun -p 3000 -o 0.0.0.0 config.ru`
+5. Update the database schema with `bundle exec rake db:migrate`
+6. Run the app with `bundle exec shotgun -p 3000 -o 0.0.0.0 config.ru`
+
+## Test
+
+1. Create a `dry_web_blog_test` Postgres database
+2. Update the database schema with `RACK_ENV=test bundle exec rake db:migrate`
+3. Run the tests with `bundle exec rake spec`
```

choonkeat replied:
is `RACK_ENV` the correct env to set?

solnic replied:
@choonkeat yes dry-web uses `RACK_ENV` to set the env

timriley replied:
Good suggestion, @choonkeat — I've just updated the README in https://github.com/dry-rb/dry-web-blog/commit/be130ae052176922ee806f036777713d133877e0

choonkeat commented on apps/main/lib/main/views/home.rb:
```diff
@@ -0,0 +1,18 @@
+      end
+
+      expose :articles do
+        articles_repo.home_page_listing
```
asking so I understand the relationship of strings, file names and variable names here
- does `"repositories.articles_repo"` do a `require_relative "../../repositories/articles_repo.rb"`?
- variable `articles_repo` name is extracted from the string? and automagically be assigned with `Main::Repositories::ArticlesRepo` class loaded from the .rb ?

solnic replied:
> does "repositories.articles_repo" do a require_relative "../../repositories/articles_repo.rb"?

No, it does `require "main/repositories/articles_repo"` because we have load paths set

> variable articles_repo name is extracted from the string? and automagically be assigned with Main::Repositories::ArticlesRepo class loaded from the .rb ?

Yes, dependency identifier is used to determine namespaces and class names

timriley replied:
And here's a few more notes about your questions, @choonkeat:

> > does "repositories.articles_repo" do a require_relative "../../repositories/articles_repo.rb"?
>
> No, it does require "main/repositories/articles_repo" because we have load paths set

It's probably worth noting that [this](https://github.com/dry-rb/dry-web-blog/blob/741609eedc59599b52e29c5aa77a30073a95721e/apps/main/system/boot.rb#L3) is the part of our app that triggers this `require` to happen. When we run `Main::Container.finalize!`, it crawls over the container's auto_register paths ([configured here](https://github.com/dry-rb/dry-web-blog/blob/741609eedc59599b52e29c5aa77a30073a95721e/apps/main/system/main/container.rb#L15)) and for each Ruby source file it finds, it `require`s that file, then makes a matching container registration that'll return an instance of the class defined in that file. So by the time this `"repositories.articles_repo"` dependency is injected here, it's already had its file required.

(FWIW, there's a different story when it comes to unit testing – in those cases, we don't finalize the container, and it will lazily load/require files for us at the first moment we try and resolve them from the container, which makes for a much faster test cycle, since it doesn't load unnecessary files.)

> variable articles_repo name is extracted from the string

Yeah, the name for the local getter can be inferred from the dependency identifier like you've seen here, but we can also be explicit in naming them:

```ruby
inclue Main::Import[repo: "persistence.articles_repo"]
```

This would give us `repo` (instead of `articles_repo`) as the local name for the dependency.

choonkeat replied:
Thanks for the explanation @solnic @timriley 🙇

choonkeat commented on lib/persistence/relations/articles.rb:
```diff
@@ -4,7 +4,7 @@ class Articles < ROM::Relation[:sql]
 
       def published
-        where(published: true)
+        by_published(true)
```
so indexed columns get a `by_*` method?

solnic replied:
Yes, rom creates a generic `by_pk` for you and for every indexed attribute you get `by_*` methods too

timriley committed:
Add README notes about migrating DB and testing

timriley committed:
Add redcarpet gem for Markdown rendering

timriley committed:
Move feature spec into shared “main/” directory

This will hold all the specs for the main sub-app

timriley committed:
Require sub-app containers in spec_helper

timriley committed:
Add struct class for Article and render body markdown text as HTML

timriley committed:
Make default article factory attributes strings, not arrays

timriley committed:
Remove stray save_and_open_page

timriley committed:
Require capybara in base spec_helper instead of web_spec_helper

This will allow us to use the DSL on plain HTML strings when we test our views

timriley committed:
Move time_match require into base spec_helper

Ready access to time manipulation is something that’s handy to have across our whole suite

timriley committed:
Add a unit test for the home view

Assert that we’re rendering markdown for each article

choonkeat commented on apps/main/lib/main/repositories/articles_repo.rb:
```diff
@@ -1,8 +1,11 @@
   module Repositories
     class ArticlesRepo < Blog::Repository[:articles]
+      struct_namespace Main::Entities
+
```
wasn't too clear what these 2 changes meant to config? (esp requiring article but declaring namespace for entities)

timriley replied:
@choonkeat Doing the `require` like this isn't _technically_ required, but I did it for 2 purposes:

1. We need to ensure the `Main::Entities` constant exists before we can refer to it as the struct namespace. Requiring `"main/entities/article"` ensures that this module has been defined.
2. Using `"main/entities/article"` as the require also kind of acts as a "hint" that this is the entity object we're likely to be returning from this repo.

So that's why I added it.

An alternative approach would've been something like this, defining the module inline.

```ruby
module Main
  module Entities
  end

  module Repositories
    class ArticlesRepo < Blog::Repository[:articles]
      struct_namespace Main::Entities
```

Or perhaps we could also create a single file with an empty `Entities` module so we could just require that instead of a particular entity, e.g.

```ruby
require "main/entities"

module Main
  module Repositories
    class ArticlesRepo < Blog::Repository[:articles]
      struct_namespace Main::Entities
```

Now that I think about it, this feels a little nicer.

What I _also_ expect to happen is that, once we want to build a second repository in the Main sub-app, I'll extract the `struct_namespace` declaration out into a shared `Main::Repository` superclass, which can be where we think about adjusting how we ensure `Main::Entities` exists.

One last thing – all of this has been because in a dry-web app, one principle we follow is that _any_ file should be able to be loaded on its own and run, without any of the rest of the system having been loaded first. So that means each file needs to be responsible for its requires and everything else.

halogenandtoast commented on apps/main/lib/main/repositories/articles_repo.rb:
```diff
@@ -0,0 +1,14 @@
+    class ArticlesRepo < Blog::Repository[:articles]
+      struct_namespace Main::Entities
+
+      def home_page_listing
```
I wonder if there is a better naming opportunity here? I would imagine that "home_page" doesn't have a direct correlation to `ArticlesRepo` and perhaps it shouldn't know anything about a home page.

timriley replied:
You're right, we could probably just call this `#listing` and accept limit as an argument.

halogenandtoast commented on apps/main/lib/main/entities/article.rb:
```diff
@@ -0,0 +1,20 @@
+module Main
+  module Entities
+    class Article < ROM::Struct
+      def body_html
```
I wonder if this could be better served by some sort of view object. Having an Entity be aware of HTML seems to be a cross cutting concern.

AMHOL replied:
Could use a [view part](http://dry-rb.org/gems/dry-view/view-parts/) for this.

timriley replied:
@AMHOL @halogenandtoast Yep, I was considering that a view part would the ideal final place for this, but I was conscious about not introducing too many things all at once in this PR.

That said, if we moved it to a view part, it means we could _remove_ this entity for now, so that's not a bad trade-off. I'll go ahead and do that :)

JuanitoFatas commented on Gemfile:
```diff
@@ -26,6 +26,7 @@ gem "dry-transaction", ">= 0.10.0"
 gem "dry-types"
 gem "dry-validation"
 gem "dry-view", ">= 0.3.0"
+gem "redcarpet" # Markdown rendering
```
fyi, not important, GitHub recently switched to use https://github.com/gjtorikian/commonmarker. ([html-pipeline](https://github.com/jch/html-pipeline)'s markdown bundled gem).

JuanitoFatas commented on db/migrate/20170711235422_create_articles.rb:
```diff
@@ -0,0 +1,14 @@
+ROM::SQL.migration do
+  change do
+    create_table :articles do
+      primary_key :id
```
Suggestion: What do you think about the framework generates primary key of `BIGINT` type?
http://www.mccartie.com/2016/12/05/rails-5.1.html

timriley replied:
@JuanitoFatas certainly a valid thing to consider – probably worth bringing up in the rom-rb or sequel projects? For this application here my plan is to just follow the typical approaches for the various gems we use.

JuanitoFatas commented on db/migrate/20170711235422_create_articles.rb:
```diff
@@ -0,0 +1,14 @@
+      column :slug, :text, null: false, unique: true, index: true
+      column :title, :text, null: false
+      column :body, :text, null: false
+      column :published, :boolean, default: false, null: false, index: true
```
Maybe we could remove this field since we have `published_at` ☺️

timriley replied:
@JuanitoFatas I was actually thinking there's still merit in having both columns. For instance, if you wanted to unpublish something but still preserve its original publishing date, you'd need both the boolean and the timestamp as separate columns. Does that seem reasonable to you?

JuanitoFatas commented on lib/persistence/relations/articles.rb:
```diff
@@ -0,0 +1,15 @@
+module Persistence
+  module Relations
```
Could I name this to `Queries` or it has to be `Relations`? 🤔

GustavoCaso replied:
I think `rom` allow you to specify the name you want. Here in this project rom `auto registration` process is set here https://github.com/dry-rb/dry-web-blog/blob/master/system/boot/rom.rb#L34

But you can pass multiple options to the `auto_register` method that allow you to modify it the way you need it 
https://github.com/rom-rb/rom/blob/master/core/spec/unit/rom/setup/auto_registration_spec.rb#L62

JuanitoFatas replied:
I see 😊  Thanks for the explanation!

JuanitoFatas commented on spec/main/unit/views/home_spec.rb:
```diff
@@ -0,0 +1,38 @@
+    let(:article) { page.find(".article:nth-of-type(2)") }
+
+    it "renders the full markdown bodies" do
+      expect(article.native.inner_html).to include "<strong>Welcome</strong> to my blog"
```
🙋 Q: Where is this `native` method coming from?

JuanitoFatas replied:
I found it's from capybara 😂 It would be nice if the framework provides `page` object available in the test suite that makes the tests more convenient.

timriley replied:
@JuanitoFatas Yeah, you're totally right, this is a bit awkward. In this particular PR, though, I'm just trying to strike a balance between implementing a basic first feature and not trying to _introduce all the things_ at the same time 😬

timriley replied:
So I might make another issue here to introduce page objects at a later stage.

JuanitoFatas commented on apps/main/lib/main/repositories/articles_repo.rb:
```diff
@@ -0,0 +1,14 @@
+module Main
+  module Repositories
+    class ArticlesRepo < Blog::Repository[:articles]
+      struct_namespace Main::Entities
```
🙋‍♂️ What is this line doing? New to dry-rb 🙏

GustavoCaso replied:
`struct_namespace` belong to rom-rb https://github.com/rom-rb/rom/blob/667dee8683e39c8f18f92e3ae295f3a4ef3f0164/repository/lib/rom/repository.rb#L75 rom returns data from the DB wrapped in a struct class.

By default if you do not specify anything it will return a struct which `class` would be `ROM::Struct::#{your relation in singular}` - for example `ROM::Struct::Article`.

What `struct_namespace` allow you is a way to specify to rom where to look for custom struct classes that will wrap the returning data - in this project they live under `apps/main/lib/main/entities/` folder https://github.com/dry-rb/dry-web-blog/blob/7f3d10ae9c963f7921471a6561ee8b05895e85b7/apps/main/lib/main/entities/article.rb

timriley committed:
Move article markdown rendering from entity to view part

timriley committed:
Change namespace for articles repo

timriley committed:
Make repo method name for article listing less view-specific

The repo doesn’t really know which views will be calling these methods, so it’s better to name its methods more generally (though still keeping them as self-descriptive as possible).

timriley committed:
Switch to CommonMarker for Markdown rendering

timriley commented:
Thanks so much everybody for your questions, comments and feedback! I think this was a really good process to go through, and I hope we can keep it up with the subsequent PRs.

I think we've got to the point now where this one can be merged in. Let me summarise what we've done here, for the record:

We've created a home page with a date-descending list of published articles.

To do this, we've put in place all the basics to establish articles as an entity within our system:

- A database migration to create the table
- A ROM relation to model the table
- A repository in the "main" sub-app for returning a list of published articles, date-descending

We've also created an "home" view controller which uses the repository to pass the list of articles to the template, with each article value being wrapped up in a view part class that adds support for rendering the body markdown as HTML.

We've added tests for each of these elements, too. Firstly, we defined a test factory for generating articles (either as simple structs or records in the database), then we added unit tests for the view and view part, as well as a single integration/feature spec to test for the correct homepage being served up via a web request.

GustavoCaso commented:
Awesome !!!!!!
